### ISSUE TITLE
rustc suggested fix fails to compile when using `cargo clippy --fix` with code containing todo!()

### ISSUE THREAD

**Issue by devslash-paul** [rust-lang/rust#93030]
I tried to run `cargo clippy --fix` with this code:

```rust
struct A{
    a: bool,
    b: bool,
}

fn main() {
    let f = true;
    A {
        a: todo!(),
        b: f
    };
}
```

I expected to see the clippy command succeed.

Instead, this happened: 
```
warning: failed to automatically apply fixes suggested by rustc to crate `tt2`

after fixes were automatically applied the compiler reported errors within these files:

  * /<elided>/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/macros/mod.rs
  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `f` in this scope
  --> src/main.rs:10:12
   |
10 |         b: f
   |            ^ help: a local variable with a similar name exists: `_f`

warning: unreachable expression
  --> src/main.rs:10:12
   |
9  |         a: todo!(),
   |            ------- any code following this expression is unreachable
10 |         b: f
   |            ^ unreachable expression
   |
   = note: `#[warn(unreachable_code)]` on by default

error: aborting due to previous error; 1 warning emitted
```

### Meta

`rustc --version --verbose`:
```
rustc 1.59.0-nightly (c14569225 2022-01-01)
binary: rustc
commit-hash: c145692254e86974941f2c92c643a23df0f13e82
commit-date: 2022-01-01
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

**Comment by camsteffen** [comment]
I can reproduce with `cargo fix` instead of `cargo clippy --fix` so clippy is not really related.

It looks like `f` is considered unused because `b: f` is unreachable due to `todo!()`. So `f` is renamed to `_f` via the `unused_variables` lint. Then `b: f` is no good.

I think ideally `unused_variables` should not fire.

**Comment by ehuss** [comment]
Thanks for the report!  Closing as a duplicate of #88402.
